core/src/main/scala/zio/jdbc/JdbcException.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2022 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio.jdbc
+
+import zio._
+
+import java.io.IOException
+import java.sql.{ Connection, ResultSetMetaData, SQLException }
+
+/**
+ * Trait to encapsule all the exceptions employed by ZIO JDBC
+ */
+sealed trait JdbcException extends Exception
```

Comment:
Question: what's the objective for the hierarchy of errors? 

Given it's meant to be use for ZIO error channel in the library API, i think the following requirements make sense:
* It should be tailored to assist with the error handing logic of a consumer of API, in this context I wonder:
  * Do we want to handle `DriverNotFound` explicitly, should we report it via `ZIO.die`

Reply:
I understand that it should clarify to the user what possible errors he must take into account and make easier
- I think that exposing more of the API gives more control to the user to decide how to handle the errors and this makes ZIO JDBC more powerful. But I think that making it easier to `ZIO.die` everything could be helpful at the application level.
- I would say so because it would mean less interference between JDBC and the user.
- I think that if all errors are direct cases of `JdbcException` we might lose the advantage of precisely categorizing errors. `ConnectionException only has one sublevel. We could do the same with `QueryException`. Again, we would be mixing errors from very different causes. What do you think?
- This could be very interesting if we were to provide the user with an easy way of specifying how he wants to treat retriable errors. For this, we might have to define a wrapper of `SQLTimeoutException`. But if this must be done by the user, in the end, he will have to decide on the retry policies, so maybe this should be in the documentation.
- All `JdbcExceptions` should wrap a `Throwable`, so the user could access the more specific SQL error. Currently, I don't think having more specific SQL errors is useful internally, but maybe in the future is helpful.

I agree that the API is more complicated. Maybe the solution is to have less categorization (e.g. using only one level of subclasses for `QueryException`). What do you think? Yet, you should be able now to do pattern matching over the corresponding error. I am sure @jdegoes will be able to give a more definitive opinion.

Reply:
> With the new API, do I need to pattern match over both FailedToConnect and ZSQLException in order to handle connection timeout?

Yes, it might be better to create a `TimeoutException` to deal with these cases. Although an user might want to treat differently the timeout if the cause was the failure to make a connection or the at the query level. Maybe it's better to have this information in the `TimeoutException`.